Mathlib/Lean/Meta/RefinedDiscrTree/Basic.lean

```diff
@@ -1,178 +1,266 @@
-  | .star ..   => 0
-  | .opaque .. => 1
-  | .const ..  => 2
-  | .fvar ..   => 3
-  | .bvar ..   => 4
-  | .lit ..    => 5
-  | .sort      => 6
-  | .lam       => 7
-  | .forall    => 8
-  | .proj ..   => 9
-
-/-- The order on `Key` used in the `RefinedDiscrTree`. -/
-private def Key.lt : Key → Key → Bool
-  | .star i₁,       .star i₂       => i₁ < i₂
-  | .const n₁ a₁,   .const n₂ a₂   => Name.quickLt n₁ n₂ || (n₁ == n₂ && a₁ < a₂)
-  | .fvar f₁ a₁,    .fvar f₂ a₂    => Name.quickLt f₁.name f₂.name || (f₁ == f₂ && a₁ < a₂)
-  | .bvar i₁ a₁,    .bvar i₂ a₂    => i₁ < i₂ || (i₁ == i₂ && a₁ < a₂)
-  | .lit v₁,        .lit v₂        => v₁ < v₂
-  | .proj s₁ i₁ a₁, .proj s₂ i₂ a₂ => Name.quickLt s₁ s₂ ||
-    (s₁ == s₂ && (i₁ < i₂ || (i₁ == i₂ && a₁ < a₂)))
-  | k₁,             k₂             => k₁.ctorIdx < k₂.ctorIdx
-
-instance : LT Key := ⟨fun a b => Key.lt a b⟩
-instance (a b : Key) : Decidable (a < b) := inferInstanceAs (Decidable (Key.lt a b))
-
 private def Key.format : Key → Format
-  | .star i                 => "*" ++ Std.format i
+  | .star id                => f!"*{id}"
   | .opaque                 => "◾"
-  | .const k a              => "⟨" ++ Std.format k ++ ", " ++ Std.format a ++ "⟩"
-  | .fvar k a               => "⟨" ++ Std.format k.name ++ ", " ++ Std.format a ++ "⟩"
-  | .lit (Literal.natVal v) => Std.format v
-  | .lit (Literal.strVal v) => repr v
-  | .sort                   => "sort"
-  | .bvar i a               => "⟨" ++ "#" ++ Std.format i ++ ", " ++ Std.format a ++ "⟩"
+  | .const name nargs       => f!"⟨{name}, {nargs}⟩"
+  | .fvar fvarId nargs      => f!"⟨{fvarId.name}, {nargs}⟩"
+  | .lit (Literal.natVal n) => f!"{n}"
+  | .lit (Literal.strVal s) => f!"{s.quote}"
+  | .sort                   => "Sort"
+  | .bvar i nargs           => f!"⟨#{i}, {nargs}⟩"
   | .lam                    => "λ"
   | .forall                 => "∀"
-  | .proj s i a             => "⟨" ++ Std.format s ++"."++ Std.format i ++", "++ Std.format a ++ "⟩"
+  | .proj name idx nargs    => f!"⟨{name}.{idx}, {nargs}⟩"
 
 instance : ToFormat Key := ⟨Key.format⟩
 
+/--
+Converts an entry (i.e., `List Key`) to the discrimination tree into
+`MessageData` that is more user-friendly.
+-/
+partial def keysAsPattern (keys : Array Key) : CoreM MessageData := do
+  let (msg, keys) ← go (paren := false) |>.run keys.toList
+  if !keys.isEmpty then
+    throwError "illegal discrimination tree entry: {keys.map Key.format}"
+  return msg
+where
+  /-- Get the next key. -/
+  next : StateRefT (List Key) CoreM Key := do
+    let key :: keys ← get | throwError "illegal discrimination tree entry: {keys.map Key.format}"
+    set keys
+    return key
+  /-- Format the application `f args`. -/
+  mkApp (f : MessageData) (nargs : Nat) (paren : Bool) : StateRefT (List Key) CoreM MessageData :=
+    if nargs == 0 then
+      return f
+    else do
+      let mut r := f
+      for _ in [:nargs] do
+        r := r ++ m!" {← go}"
+      return parenthesize m!"{r}" paren
```

Comment:
```suggestion
      let args ← (List.range nargs).mapM fun _ => go
      return parenthesize (MessageData.joinSep (f :: args) " ") paren
```
though you might want to consider using `Format.line` instead of `" "`